capstone/src/main/java/com/google/sps/servlets/GroupPostDataServlet.java

```diff
@@ -21,50 +22,89 @@
+import java.util.Map;
+import java.util.HashSet;
 import java.io.PrintWriter;
 import java.net.MalformedURLException;
 import java.net.URL;
 import com.google.sps.Objects.User;
 import com.google.sps.Objects.Post;
 import com.google.sps.Objects.Comment;
+import com.google.sps.Objects.PostResponse;
 
 @WebServlet("/group-post")
 public class GroupPostDataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
+    String userId = "";
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      userId = userService.getCurrentUser().getUserId();
+    }
+
     Query query = new Query("Post").addSort("timestamp", SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
     List<Post> posts = new ArrayList<>();
+    List<Long> likedPosts = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
       posts.add(getPostEntity(entity));
+      ArrayList<String> likes = (ArrayList<String>) entity.getProperty("likes");
+      if(likes != null && likes.contains(userId)) {
+        likedPosts.add(entity.getKey().getId());
+      }
     }
 
     // Convert to json
+    PostResponse postsRes = new PostResponse(posts, likedPosts, userId);
     response.setContentType("application/json;");
-    response.getWriter().println(new Gson().toJson(posts));
+    response.getWriter().println(new Gson().toJson(postsRes));
   }
 
-  public Post getPostEntity(Entity entity) {
+  private Post getPostEntity(Entity entity) {
+    long postId = entity.getKey().getId();
     long timestamp = (long) entity.getProperty("timestamp");
     String authorId = (String) entity.getProperty("authorId");
     String postText = (String) entity.getProperty("postText");
     String challengeName = (String) entity.getProperty("challengeName");
     String img = (String) entity.getProperty("img");
-    HashSet<String> likes = (entity.getProperty("likes") == null) ? new HashSet<>() : new HashSet<String>((ArrayList<String>) entity.getProperty("likes"));
-    ArrayList<Comment> comments = (ArrayList<Comment>) entity.getProperty("comments");
-    Post userPost = new Post(authorId, postText, comments, challengeName, timestamp, img, likes);
+    HashSet<String> likes = (entity.getProperty("likes") == null) 
+      ? new HashSet<>() 
+      : new HashSet<String>((ArrayList<String>) entity.getProperty("likes"));   
+    ArrayList<Comment> comments = new ArrayList<>();
+  
+    // Create list of comment objects from list of comment entities 
+    if (entity.getProperty("comments") != null) {
+      createCommentObjectList(comments, entity);
+    }
+
+    Post userPost = new Post(postId, authorId, postText, comments, challengeName, timestamp, img, likes);
     return userPost;
   }
```

Comment:
Would it make more sense to define this in the Post.java file as a public `fromEntity(Entity entity)` function?

Reply:
I have all my create/post entity functions in the servlet files, so I think I would need to move all those functions to their respective classes to be consistent. Are there any drawbacks to leaving the entity portion in the servlet file?

Reply:
nvm will make this change! makes sense to move the post object creation to Post.java.